engine/src/main/java/org/hibernate/validator/spi/cascading/ValueExtractor.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.spi.cascading;
+
+public interface ValueExtractor<T> {
+
+	void extractValues(T originalValue, ValueReceiver receiver);
+
+	interface ValueReceiver {
+
+		void objectValue(Object object);
+
+		void listValue(int i, Object object);
```

Comment:
Wondering if `listItem` and `iterableItem` (or `listElement` and `iterableElement`) would be more descriptive? It's just a detail but I was surprised to see an index for `listValue` and then I understood it was only an item, not the list.